src/Npgsql/VolatileResourceManager.cs

```diff
@@ -224,6 +232,38 @@ public void InDoubt(Enlistment enlistment)
             }
         }
 
+        public void WaitIfRequired()
+        {
+            try
+            {
+                if (_secondPhaseEndedOrNotNeedingWait || _isRMCall.Value ||
+                    _transaction?.TransactionInformation.Status == System.Transactions.TransactionStatus.Active)
+                    return;
```

Comment:
It is not an optimization, but I should simplify this by using a `ManualResetEventSlim` instead.